src/operator/join.rs

```diff
@@ -0,0 +1,217 @@
+        Circuit, Scope, Stream,
+    },
+};
+use std::{borrow::Cow, marker::PhantomData};
+
+impl<P, I1> Stream<Circuit<P>, I1>
+where
+    P: Clone + 'static,
+{
+    /// Apply [`Join`] operator to `self` and `other`.
+    ///
+    /// See [`Join`] operator for more info.
+    pub fn join<K, V1, V2, V, W, F, I2, Z>(
+        &self,
+        other: &Stream<Circuit<P>, I2>,
+        f: F,
+    ) -> Option<Stream<Circuit<P>, Z>>
+    where
+        K: KeyProperties,
+        V1: KeyProperties,
+        V2: KeyProperties,
+        W: ZRingValue,
+        I1: IndexedZSet<K, V1, W>,
+        for<'a> &'a I1: IntoIterator<Item = (&'a K, &'a I1::ZSet)>,
+        for<'a> &'a I1::ZSet: IntoIterator<Item = (&'a V1, &'a W)>,
+        I2: IndexedZSet<K, V2, W>,
+        for<'a> &'a I2: IntoIterator<Item = (&'a K, &'a I2::ZSet)>,
+        for<'a> &'a I2::ZSet: IntoIterator<Item = (&'a V2, &'a W)>,
+        F: Fn(&K, &V1, &V2) -> V + 'static,
+        V: 'static,
+        Z: Clone + MapBuilder<V, W> + 'static,
+    {
+        self.circuit()
+            .add_binary_operator(Join::new(f), self, other)
+    }
+}
+
+/// Join two indexed Z-sets.
+///
+/// The operator takes two streams of indexed Z-sets and outputs
+/// a stream obtained by joining each pair of inputs.
+///
+/// An indexed Z-set is a map from keys to a Z-set of values associated
+/// with each key.  Both input streams must use the same key type `K`.
+/// Indexed Z-sets are produced for example by the
+/// [`Index`](`crate::operator::Index`) operator.
+///
+/// # Type arguments
+///
+/// * `K` - key type.
+/// * `V1` - value type in the first input stream.
+/// * `V2` - value type in the second input stream.
+/// * `V` - value type in the output Z-set.
+/// * `W` - weight type used by both inputs.
+/// * `F` - join function type: maps key and a pair of values from input Z-sets
+///   to an output value.
+/// * `I1` - indexed Z-set type in the first input stream.
+/// * `I1` - indexed Z-set type in the second input stream.
+/// * `Z` - output Z-set type.
+pub struct Join<K, V1, V2, V, W, F, I1, I2, Z> {
+    join_func: F,
+    _val_types: PhantomData<(K, V1, V2, V, W)>,
+    _container_types: PhantomData<(I1, I2, Z)>,
+}
+
+impl<K, V1, V2, V, W, F, I1, I2, Z> Join<K, V1, V2, V, W, F, I1, I2, Z> {
+    pub fn new(join_func: F) -> Self {
+        Self {
+            join_func,
+            _val_types: PhantomData,
+            _container_types: PhantomData,
+        }
+    }
+}
+
+impl<K, V1, V2, V, W, F, I1, I2, Z> Operator for Join<K, V1, V2, V, W, F, I1, I2, Z>
+where
+    K: 'static,
+    V1: 'static,
+    V2: 'static,
+    W: 'static,
+    I1: 'static,
+    I2: 'static,
+    F: 'static,
+    V: 'static,
+    Z: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Join")
+    }
+    fn clock_start(&mut self, _scope: Scope) {}
+    fn clock_end(&mut self, _scope: Scope) {}
+}
+
+fn join_inner<K, V1, V2, V, W, F, I1, I2, Z>(i1: &I1, i2: &I2, join_func: &F) -> Z
```

Comment:
Small nit, we probably want to join *into* a pre-existing output index instead of creating our own

Reply:
I'm talking about allowing reusing physical allocations, just making `&mut Z` a parameter on the function

Reply:
I can refactor the function that way, but we wouldn't be able to take advantage of this in our current setup.  Where would a Z-set to reuse come from? The set produced at the previous iteration has been consumed by downstream operators...